core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java

```diff
@@ -167,39 +158,54 @@
     private static final String CONFIG_ROOTS_LIST = "META-INF/quarkus-config-roots.list";
     private static final String[] NO_STRINGS = new String[0];
 
+    static final int DEFAULT_CONVERTER_PRIORITY = 200;
+
     public ConfigurationSetup() {
     }
 
     @BuildStep
-    public void setUpConverters(BuildProducer<ConfigurationCustomConverterBuildItem> configurationTypes) {
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                InetSocketAddress.class,
-                InetSocketAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                CidrAddress.class,
-                CidrAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                InetAddress.class,
-                InetAddressConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Pattern.class,
-                RegexConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                CipherSuiteSelector.class,
-                CipherSuiteSelectorConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Protocol.class,
-                ProtocolConverter.class));
-        configurationTypes.produce(new ConfigurationCustomConverterBuildItem(
-                200,
-                Path.class,
-                PathConverter.class));
+    public void setUpConverters(BuildProducer<ConfigurationCustomConverterBuildItem> converters) {
+        for (final Converter<?> converter : ServiceLoader.load(Converter.class)) {
```

Comment:
@dmlloyd Thank you for inspecting this code. Before we move forward I would like to discuss some caveats I see in regards to this idea. I may be wrong since I'm not very familiar with the SubstrateVM, but after looking at [this code](https://github.com/cstancu/native-image-service-loader-demo), specifically this [reflections config](https://github.com/cstancu/native-image-service-loader-demo/blob/master/reflection_config.json), I assume that there is a need to produce `ReflectiveClassBuildItem` for every `Converter` implementation we have in our services file. This in turn must be done in `@BuildStep` thus leading us back to the deployment time `Converter` iteration. Therefore it seems impossible to me to move this to the runtime because even when we do it, we still have to have `ReflectiveClassBuildItem` produced in deployment time, which is charged to the same problem you pointed out in your comment.

Reply:
This is not exactly correct.  Reflection actually works completely fine in native images, as long as you have the reflection object during static initialization.  The reflective class build items are not as magical as they seem: really all they do is create the reflection objects during static initialization so that they're available at run time.  By creating the `Converter` instances during static initialization, even the ones which use reflection, we basically bypass this entire process and no reflective build items are needed.

Reply:
Ok, I get it now. So if reflections are allowed during static initialization, my idea is the following.

Modify synthetic `io/quarkus/runtime/generated/RunTimeConfig.class` by adding `static final` field whitch will hold a list of all `Converter` impls, something like:

```java
private static final List<ConverterItem<?>> CONVERTERS = ConverterSupport.getConverters();
```

And have a `ConverterSupport` class like this:

```java
public class ConverterSupport {

  static final int DEFAULT_CONVERTER_PRIORITY = 200;

  static int getConverterPriority(final Class<? extends Converter<?>> converterClass) {
    return Optional
      .ofNullable(converterClass.getAnnotation(Priority.class))
      .map(Priority::value)
      .orElse(DEFAULT_CONVERTER_PRIORITY);
  }

  private static <T> ConverterItem<?> converterToItem(final Converter<T> converter) {
    final Class<? extends Converter<T>> converterClass = (Class<? extends Converter<T>>) converter.getClass();
    final Class<T> convertedType = ConverterFactory.getConverterType(converter);
    final int priority = getConverterPriority(converterClass);
    return new ConverterItem<T>(convertedType, converter, priority);
  }

  public static List<ConverterItem<?>> getConverters() {
    return StreamSupport
      .stream(ServiceLoader.load(Converter.class).spliterator(), false)
      .map(ConverterSupport::converterToItem)
      .collect(Collectors.toList());
  }
}
```

And modify synthetic `io/quarkus/runtime/generated/RunTimeConfig.class` again to change `getRunTimeConfiguration` method by adding `for/each` loop which will iterate over items in `CONVERTERS` list and populate `SmallRyeConfigBuilder`:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  for (ConverterItem item : CONVERTERS) {
    builder.withConverter(item.getConvertedType(), item.getPriority(), item.getConverter());
  }
  // some other stuff
}
```

Instead of:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  builder.withConverter(SomeType1.class, 200, new SomeTypeConverter1());
  builder.withConverter(SomeType2.class, 200, new SomeTypeConverter2());
  builder.withConverter(SomeType3.class, 200, new SomeTypeConverter3());
  builder.withConverter(SomeType4.class, 200, new SomeTypeConverter4());
  // some other stuff
}
```

But I'm completely green when it comes to Gizmo :( I have no idea where to start, i.e. how to generate a basic foor loop. I once asked about some Gizmo related materials (tutorials, demos, examples, etc) on Zulip but since there was no response I assumed that such goodies are not available.

Reply:
Ok, I have another idea. Much simpler and doable with my level of Gizmo experience :) The `ConverterSupport` class will look like this:

```java
public class ConverterSupport {

  private static final List<ConverterItem<?>> CONVERTERS = getConverters();
  private static final int DEFAULT_CONVERTER_PRIORITY = 200;

  static int getConverterPriority(final Class<? extends Converter<?>> converterClass) {
    return Optional
      .ofNullable(converterClass.getAnnotation(Priority.class))
      .map(Priority::value)
      .orElse(DEFAULT_CONVERTER_PRIORITY);
  }

  private static <T> ConverterItem<?> converterToItem(final Converter<T> converter) {
    final Class<? extends Converter<T>> converterClass = (Class<? extends Converter<T>>) converter.getClass();
    final Class<T> convertedType = ConverterFactory.getConverterType(converter);
    final int priority = getConverterPriority(converterClass);
    return new ConverterItem<T>(convertedType, converter, priority);
  }

  private static List<ConverterItem<?>> getConverters() {
    return StreamSupport
      .stream(ServiceLoader.load(Converter.class).spliterator(), false)
      .map(ConverterSupport::converterToItem)
      .collect(Collectors.toList());
  }

  public static void populate(final SmallRyeConfigBuilder builder) {
    CONVERTERS.forEach(addConverterTo(builder));
  }

  private static <T> Consumer<ConverterItem<T>> addConverterTo(final SmallRyeConfigBuilder builder) {
    return item -> builder.withConverter(
      item.getConvertedType(),
      item.getPriority(),
      item.getConverter());
  }
}
```

And in `io/quarkus/runtime/generated/RunTimeConfig.class` it will be used as:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  ConverterSupport.populate(builder);
  // some other stuff
}
```

I think invoking single static method with one argument is simpler than generating `for/each` loop. 

What do you think?

Reply:
Additionally we can also have a `@Template` which will be invoked in a `@Record(STATIC_INIT)` `@BuildStep` and initialize the `ConverterSupport` class.

Reply:
@dmlloyd, Can you please comment?

Reply:
It seems reasonable.  You don't have to initialize the `ConverterSupport` class because the `native-image` tool initializes all classes for you.

Reply:
Ok, this is a good news :) I will start modifying the code and will come back to you soon.

Reply:
Hi @dmlloyd 

Please take a look again. Some changes worth noting:

* `ConfigurationCustomConverterBuildItem` is no more. With new mechanism it became obsolete and I removed it.
* Updated documentation.